src/Dapr.Messaging/PublishSubscribe/DaprPublishSubscribeGrpcClient.cs

```diff
@@ -24,71 +24,37 @@ internal sealed class DaprPublishSubscribeGrpcClient : DaprPublishSubscribeClien
+    /// <param name="messageHandler">The delegate reflecting the action to take upon messages received by the subscription.</param>
     /// <param name="cancellationToken">Cancellation token.</param>
-    /// <returns>An <see cref="IAsyncEnumerable{TopicMessage}"/> containing the various messages returned by the subscription.</returns>
-    public override IAsyncEnumerable<TopicMessage> SubscribeAsync(string pubsubName, string topicName, DaprSubscriptionOptions options,
-        CancellationToken cancellationToken)
-    {
-        var receiver = new PublishSubscribeReceiver(pubsubName, topicName, options, daprClient);
-        clients[(pubsubName, topicName)] = receiver;
-
-        return receiver.SubscribeAsync(cancellationToken);
-    }
-
-    /// <summary>
-    /// Used to acknowledge receipt of a message and indicate how the Dapr sidecar should handle it.
-    /// </summary>
-    /// <param name="pubsubName">The name of the Publish/Subscribe component.</param>
-    /// <param name="topicName">The name of the topic to subscribe to.</param>
-    /// <param name="messageId">The identifier of the message to apply the action to.</param>
-    /// <param name="messageAction">Indicates the action to perform on the message.</param>
-    /// <param name="cancellationToken">Cancellation token.</param>
-    public override async Task AcknowledgeMessageAsync(string pubsubName, string topicName, string messageId,
-        TopicMessageAction messageAction, CancellationToken cancellationToken)
-    {
-        if (!clients.TryGetValue((pubsubName, topicName), out var receiver))
-        {
-            throw new Exception($"Unable to find receiver instance for specified publish/subscribe component name '{pubsubName}' and topic '{topicName}'.");
-        }
-
-        await receiver.AcknowledgeMessageAsync(messageId, messageAction, cancellationToken);
-    }
-
-    /// <summary>
-    /// Unsubscribes a streaming subscription for the specified Publish/Subscribe component and topic.
-    /// </summary>
-    /// <param name="pubsubName">The name of the Publish/Subscribe component.</param>
-    /// <param name="topicName">The name of the topic to subscribe to.</param>
-    /// <param name="cancellationToken">Cancellation token.</param>
-    public override async Task UnsubscribeAsync(string pubsubName, string topicName, CancellationToken cancellationToken)
+    /// <returns></returns>
+    public override IAsyncDisposable Register(string pubSubName, string topicName, DaprSubscriptionOptions options, TopicMessageHandler messageHandler, CancellationToken cancellationToken)
     {
-        if (!clients.TryGetValue((pubsubName, topicName), out var receiver))
-        {
-            throw new Exception($"Unable to find receiver instance for specified publish/subscribe component name '{pubsubName}' and topic '{topicName}'.");
-        }
+        var key = (pubSubName, topicName);
+        if (clients.ContainsKey(key))
```

Comment:
There's a race condition here if two threads call `Register()` at the same time.  It could be resolved with a simple `lock` or, say, a `ConcurrentDictionary` and `Lazy<T>` scheme.

Reply:
Done